src/main/kotlin/com/atlassian/performance/tools/report/api/NonparametricFacade.kt

```diff
@@ -0,0 +1,58 @@
+package com.atlassian.performance.tools.report.api
+
+import com.numericalmethod.suanshu.stats.test.rank.wilcoxon.WilcoxonRankSum
+import org.apache.commons.math3.stat.descriptive.rank.Median
+import org.apache.commons.math3.stat.inference.KolmogorovSmirnovTest
+import java.lang.Exception
+
+class NonparametricFacade(
+    private val baseline: DoubleArray,
+    private val experiment: DoubleArray,
+    private val mwAlpha: Float = 0.05F,
+    private val ksAlpha: Float = 0.05F
+) {
+
+    private val shift: Double by lazy { hodgesLehmannDistance(baseline, experiment) }
+    private val baselineMedian: Double by lazy { Median().evaluate(baseline) }
+
+    /**
+     * Make sure equalDistributionsAfterShift is TRUE before accessing this property
+     */
+    val locationShift: Double by lazy {
+        if (!equalDistributionsAfterShift) throw Exception("Distribution shapes are different at $ksAlpha significance level. Location shift is meaningless.")
+        return@lazy shift
+    }
+
+    /**
+     * Location shift as a percentage of baseline's median
+     */
+    val percentageShift: Float by lazy { (shift / baselineMedian).toFloat() }
+
+    /**
+     * Tests for equality of probability distribution shapes ignoring location
+     */
+    val equalDistributionsAfterShift: Boolean by lazy {
+        val shiftedExperiment = experiment.copyOf()
+        shiftedExperiment.indices.forEach { shiftedExperiment[it] += shift }
```

Comment:
Alternative, indexing-free impl:
```kotlin
val shiftedExperiment = experiment
    .map { it + shift }
    .toDoubleArray()
```

Reply:
Wouldn't this trigger autoboxing/unboxing and be less efficient performance-wise?

Reply:
Not until proven with a profiler

Reply:
I don't get it. My initial way it just iterates over an array and increments its values. It's "direct on the metal" - it's a very efficient operation.
The way you suggested, it autoboxes each double, then it populates a collection (which probably is resized a few times) and then it converts this collection to an array.

How can the 2nd way be as fast as the 1st?!

Reply:
So I made this experiment:
```
        val sample = NormalDistribution(100.0, 10.0).sample(10_000_000)
        val shift = 3.0
        val op1 = measureTimeMillis {
            val copy1 = sample.copyOf()
            copy1.indices.forEach { copy1[it] += shift }
        }
        println(op1)
        val op2 = measureTimeMillis {
            val copy2 = sample
                .map { it + shift }
                .toDoubleArray()
        }
        println(op2)
```
The output is:
```
73
3640
```
The suggested way is 50 times slower!
I'm sorry, but I'm reverting to the initial implementation :)

Reply:
http://wiki.c2.com/?PrematureOptimization
You were optimizing a piece of code without even having a performance problem and assumed it is faster without testing it.
Now you proved it with a microbenchmark, so it's a different story.

Of course, microbenchmarks tend to be artificially bloated to highlight the difference. In real use cases do we have 10 million sample size? And if we had, would this array copying the bottleneck or would we crash and burn somewhere else beforehand (perhaps parsing? or holding 10 million action metrics in memory throwing OOMEs?). It's possible that this optimization yields a grand total of zero detectable improvement for the actual use cases, we still don't know.

Sure we can keep this micro-optimization, but we should keep in mind how wasteful this can be. Both in terms of dev time and code readability cost.

Reply:
> You were optimizing a piece of code without even having a performance problem and assumed it is faster without testing it.

But I wasn't really optimizing anything... I just wrote it with a natural assumption that we should prefer primitives over Objects whenever possible.
https://stackoverflow.com/a/5199425
https://www.baeldung.com/java-primitives-vs-objects
Why is it bad? Shouldn't we always keep that in mind and write the code based on that assumption?

Reply:
We absolutely should not prefer primitives over Objects. We'd have to use arrays everywhere. Objects let us use collections like lists, sets, maps. Objects let us use streams. Arrays are only useful for low-level implementations and optimizations, which are rarely needed.

The microbenchmarks you linked again suffer from the same fundamental flaw as your microbenchmark. When you have a loop of 4 billion iterations, autoboxing is the least of your problems. You most likely have a way more expensive operation. It's simply an unrealistic scenario.